### PR TITLE
Refactor `thrust::[stable_]partition[_copy]` to use `cub::DevicePartition`

### PR DESCRIPTION
**Pull request by elstehle**
## Description

Closes https://github.com/NVIDIA/cccl/issues/1397
Closes https://github.com/NVIDIA/cccl/issues/1383

**Changes:**
- Adds option to take a two distinct output iterators in `AgentSelectIf`: (1) one for the selected items, (2) one for the rejected items in 
  - This is required to implement `thrust::[stable_]partition_copy` interfaces 
- Adds dynamic 32/64-bit offset type-dispatch to `thrust::[stable_]partition[_if]`
- Adds tests for large number of items for  `thrust::stable_partition_if`
- Fixes an offset computation when writing `rejected` items for 64-bit offset types

**Follow-On Tasks** 
The following tasks are left for follow-on PRs:
- Support large number of items in cub's `cub::DevicePartition` interfaces is left to https://github.com/NVIDIA/cccl/issues/1437
- Adding the interfaces that are exposed in thrust (i.e., "stencil_if", partitioning selected items to a different iterator than rejected items) is deferred to https://github.com/NVIDIA/cccl/issues/1436

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: thrust/thrust/system/cuda/detail/partition.h [hunk 1]
```diff
@@ -38,1041 +38,349 @@
...
+  execution_policy<Derived>& policy,
+  InputIt first,
+  InputIt last,
+  StencilIt stencil,
+  OutputIt output,
+  Predicate predicate)
+{
+  using size_type = typename iterator_traits<InputIt>::difference_type;
+
+  size_type num_items = static_cast<size_type>(thrust::distance(first, last));
```
Comment (gevtushenko) [inline thrust/thrust/system/cuda/detail/partition.h:174]: question: I think distance is guaranteed to return difference type. Do we need this cast?

File: thrust/thrust/system/cuda/detail/partition.h [hunk 2]
```diff
@@ -38,1041 +38,349 @@
...
-    return first + num_selected;
-  }
-}    // namespace __partition
+template <typename Derived, typename InputIt, typename StencilIt, typename Predicate>
+THRUST_RUNTIME_FUNCTION InputIt inplace_partition(
+  execution_policy<Derived>& policy, InputIt first, InputIt last, StencilIt stencil, Predicate predicate)
+{
+  // Element type of the input iterator
+  using value_t         = typename iterator_traits<InputIt>::value_type;
+  std::size_t num_items = static_cast<std::size_t>(thrust::distance(first, last));
```
Comment (gevtushenko) [inline thrust/thrust/system/cuda/detail/partition.h:238]: question: I'm not sure about this cast. Temporary array checks for `num_items > 0`. If user accidentally provides last < first, we'll try to allocate size_t max. Do you think leaving num items signed until we reach dispatch is any better?

File: thrust/thrust/system/cuda/detail/partition.h [hunk 3]
```diff
@@ -38,1041 +38,349 @@
...
-
-    int tile_size = partition_plan.items_per_tile;
-    size_t num_tiles = cub::DivideAndRoundUp(num_items, tile_size);
-
-    size_t vshmem_storage = core::vshmem_size(partition_plan.shared_memory_size,
-                                              num_tiles);
-
-    cudaError_t status = cudaSuccess;
+    // Return for empty problems
     if (num_items == 0)
```
Comment (gevtushenko) [inline thrust/thrust/system/cuda/detail/partition.h:123]: same question: maybe worth rewriting as > 0?

**Review by gevtushenko — APPROVED** [review]
This is an elegant solution enabling new functionality with little changes in CUB internals. Great work!

File: cub/cub/agent/agent_select_if.cuh [hunk 4]
```diff
@@ -791,7 +857,7 @@ struct AgentSelectIf
         OffsetT num_tile_selections     = prefix_op.GetBlockAggregate();
         OffsetT num_selections          = prefix_op.GetInclusivePrefix();
         OffsetT num_selections_prefix   = prefix_op.GetExclusivePrefix();
-        OffsetT num_rejected_prefix     = (tile_idx * TILE_ITEMS) - num_selections_prefix;
+        OffsetT num_rejected_prefix     = tile_offset - num_selections_prefix;
```
Comment (gevtushenko) [inline cub/cub/agent/agent_select_if.cuh:860]: praise: good catch!

File: thrust/thrust/system/cuda/detail/partition.h (hunk 2 above)
Comment (elstehle) [reply thrust/thrust/system/cuda/detail/partition.h:246]: Thanks, good catch on the negative ranges!

I've decided to check and return early for negative/empty ranges. I was concerned about other inconvenient side effects from negative `num_items`, like allocating the temp array with negative size for the `inplace` version, and returning the rejected iterator negative by the num items count (i.e., `rejected_result + num_items - num_selected`). So, instead we now have:
```
  if(thrust::distance(first, last) <= 0){
    return thrust::make_pair(selected_result, rejected_result);
  }
```
and 
```
  if(thrust::distance(first, last) <= 0){
    return first;
  }
```

**Review by miscco — APPROVED** [review]

**Pull request merged**